extensions/ipynb/src/notebookImagePaste.ts

```diff
@@ -0,0 +1,145 @@
+
+	async provideDocumentPasteEdits(
+		_document: vscode.TextDocument,
+		_ranges: readonly vscode.Range[],
+		dataTransfer: vscode.DataTransfer,
+		_token: vscode.CancellationToken
+	): Promise<vscode.DocumentPasteEdit | undefined> {
+
+		// get foundational notebook data
+		const activeNotebook = vscode.window.activeNotebookEditor?.notebook;
+		const activeSelection = vscode.window.activeNotebookEditor?.selection;
+		const currentCell = activeNotebook?.getCells(activeSelection);
+		if (!currentCell || currentCell?.length !== 1) {
+			return undefined;
+		}
+		const initialAttachments = currentCell[0].metadata?.custom?.attachments;
+		const initialize = !initialAttachments ? true : false;
+
+		// get b64 data from paste
+		const dataItem = dataTransfer.get('image/png');
+		if (!dataItem) {
+			return undefined;
+		}
+		const fileDataAsUint8 = await dataItem.asFile()?.data();
+		if (!fileDataAsUint8) {
+			return undefined;
+		}
+		const b64string = encodeBase64(fileDataAsUint8);
+
+		// get filename data from paste
+		let pasteFilename = dataItem.asFile()?.name;
+		const filePrefix = pasteFilename?.split('.')[0];
+		const fileSuffix = pasteFilename?.split('.')[1];
+		if (!filePrefix || !fileSuffix) {
+			return undefined;
+		}
+
+		// create updated metadata for cell (prep for WorkspaceEdit)
+		if (!initialize) {
+			let appendValue = 2;
+			let attachmentFilename: keyof typeof initialAttachments;
+			const sortedFilenames = sortFilenames(initialAttachments);
+			for (attachmentFilename of sortedFilenames) {
+				if (attachmentFilename === pasteFilename) {
+					const objEntries = Object.entries(initialAttachments[pasteFilename]);
+					if (objEntries.length) { // check that mime:b64 are present
+						const [, attachmentb64] = objEntries[0];
+						if (attachmentb64 !== b64string) {	// append a "-#" here. same name, diff data. this matches jupyter behavior
+							pasteFilename = filePrefix.concat(`-${appendValue++}.`) + fileSuffix;
+						}
+					}
+				}
+			}
+		}
+		if (!pasteFilename) {
+			return undefined;
+		}
+		const updatedMetadata = { 'custom': { 'attachments': { [pasteFilename]: { 'image/png': b64string } } } };
+		if (!initialize) {
+			let attachmentFilename: keyof typeof initialAttachments;
+			for (attachmentFilename in initialAttachments) {
+				updatedMetadata.custom.attachments[attachmentFilename] = initialAttachments[attachmentFilename];
+			}
+		}
+
+		// create WorkspaceEdit and apply to active notebook
+		let newCell;
+		if (updatedMetadata) {
+			newCell = vscode.NotebookEdit.updateCellMetadata(currentCell[0].index, updatedMetadata);
+		}
+		const edit = new vscode.WorkspaceEdit();
+		if (activeNotebook?.uri && newCell) {
+			edit.set(activeNotebook.uri, [newCell]);
+		}
+		vscode.workspace.applyEdit(edit);
+
+		// Build a snippet to paste
+		const snippet = new vscode.SnippetString();
+		snippet.appendText(`![${pasteFilename}](attachment:${pasteFilename})`);
```

Comment:
Should the text of image link be a snippet placeholder instead of the filename?

Reply:
Was trying to mimic the behavior of jupyter notebooks. When pasting any image, whether screenshot or file, the alt-text in square brackets is the filename itself. 

That's an easy change though if we want? Let me know.

Reply:
How about including the filename as a placeholder? That way we encourage users to add a more descriptive title while also allowing them to tab past this if they don't want

Also should the placeholder then be the original file name instead of the potentially modified filename?

Reply:
The behavior in jupyter has the placeholder being the potentially modified filename.
<img width="310" alt="image" src="https://user-images.githubusercontent.com/12552271/182461992-76744d62-09d5-4ca2-be94-386579f06b90.png">

I can change this within our implementation to have it be the original un-appended filename if you think that works best. Not quite sure what you mean by including the filename as a placeholder? Maybe there's functionality that I'm not aware of but I don't know how to implement what I think you're asking for.

Reply:
Snippets can define placeholder values that users can tab through to change if they want: https://code.visualstudio.com/docs/editor/userdefinedsnippets#_create-your-own-snippets

You can use `SnippetString.appendPlaceholder` to safely create these